### ISSUE TITLE
pnpm link unavailable package creates empty folder

### ISSUE THREAD

**Issue by paulpflug** [pnpm/pnpm#843]
hey,

I really enjoy working with pnpm. You did a great job!

I found a little bug:
```sh
# in some folder which should use somePkg
pnpm link somePkg
```
will create a empty `somePkg` folder in `pnpm-global/node-modules`, when `somePkg` isn't linked yet.

When linking the package up:
```sh
# in somePkg folder
pnpm link
```
pnpm fails with `ERROR EINVAL: invalid argument, readlink`
because it tries to use readlink on a folder.

I would suggest, that no empty folder should be created in the first place?

**Comment by zkochan** [comment]
I currently work on `pnpm list`. If you want this one fixed faster (seems easy), you can try to fix it in supi (the installation engine was moved there recently, part of an effort to divide pnpm into smaller, simpler parts).

https://github.com/pnpm/supi/blob/master/src/api/link.ts

Hm, probably the linking logic should be moved into a separate package as well.

**Comment by paulpflug** [comment]
I would add a check if `linkedPkgPath` exists [here](https://github.com/pnpm/supi/blob/master/src/api/link.ts#L58)

something like:
```js
fs = require('mz/fs')
stats = await fs.stat(linkedPkgPath)
if (!stats.isDirectory()) { throw new Error(pkgName + " not found in global path") }
```

**Comment by zkochan** [comment]
I wonder, what creates the empty folder? I think the empty folder should not be created. In that case, this line will fail https://github.com/pnpm/supi/blob/master/src/api/link.ts#L44

**Comment by paulpflug** [comment]
I think this is the trace to folder creation:
https://github.com/pnpm/supi/blob/master/src/api/getContext.ts#L80
https://github.com/pnpm/supi/blob/master/src/api/install.ts#L121
https://github.com/pnpm/supi/blob/master/src/api/install.ts#L112
https://github.com/pnpm/supi/blob/master/src/api/link.ts#L26
https://github.com/pnpm/supi/blob/master/src/api/link.ts#L59

you could probably move
https://github.com/pnpm/supi/blob/master/src/api/install.ts#L124

`if (!ctx.pkg) throw new Error('No package.json found')`

into:
https://github.com/pnpm/supi/blob/master/src/api/getContext.ts#L75-L81

```js
const files = await Promise.all([
    (opts.global ? readGlobalPkgJson(opts.prefix) : readPkgFromDir(opts.prefix)),
    readShrinkwrap(root, shrOpts),
    readPrivateShrinkwrap(root, shrOpts),
    readStore(storePath),
  ]).then((files) => {
    if (!files[0]) throw new Error('No package.json found')
    return Promise.all(files.push(mkdirp(storePath))
  })
```

**Comment by paulpflug** [comment]
or even better delay everything until package.json is loaded:
```js
const files = await (opts.global ? readGlobalPkgJson(opts.prefix) : readPkgFromDir(opts.prefix)).then( pkg => {
  if (!pkg) throw new Error('No package.json found')
  return Promise.all([
    pkg,
    readShrinkwrap(root, shrOpts),
    readPrivateShrinkwrap(root, shrOpts),
    readStore(storePath),
    mkdirp(storePath),
  ])
}) 
```

**Comment by zkochan** [comment]
I figured out what causes this... locking. Projects are locked during installation, and when there is no project folder, the lock creates one

**Comment by zkochan** [comment]
Landed in [v1.5.3](https://github.com/pnpm/pnpm/releases/tag/v1.5.3)
